velox/functions/prestosql/aggregates/MinMaxAggregates.cpp

```diff
@@ -886,18 +886,56 @@ class MinMaxNAggregateBase : public exec::Aggregate {
   DecodedVector decodedIntermediates_;
 };
 
+/// We have to use custom comparators to ensure compatibility with Presto Java.
+/// Currently Presto Java treats (+/- Nan) > Infinity, however C++ treats NaN's
+/// as nulls and moves them to end of the list if its +NaN and beginning of list
+/// if its -NaN.
 template <typename T>
-class MinNAggregate : public MinMaxNAggregateBase<T, std::less<T>> {
+struct compareLess {
+  bool operator()(const T& lhs, const T& rhs) {
+    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
+      if (std::isnan(lhs)) {
```

Comment:
If we decide to go this route, we need to do it more efficiently and systematically.  The following two functions should be kept separately (e.g. in `PrestoComparisonUtil`) and used wherever applies:
1. First we need to normalize all `-NaN` to `NaN` in input with a global function
2. Optimize the comparison
```c++
if constexpr (std::is_same_v<T, double>) {
  return reinterpret_cast<int64_t>(lhs) < reinterpret_cast<int64_t>(rhs);
}
if constexpr (std::is_same_v<T, float>) {
  return reinterpret_cast<int32_t>(lhs) < reinterpret_cast<int32_t>(rhs);
}
```

Reply:
I can move these to a common util class, one thing I am not so sure is reinterpreting doubles as int64_t , is that always correct ? Does it work no matter the values ?

Reply:
Yes it works for everything except -NaN values, which we will normalize out